File: lib/java/src/main/java/io/github/msgpact/internal/binary/ConstructBinaryEncoding.java
package io.github.msgpact.internal.binary;

import static io.github.msgpact.internal.binary.CreateChecksum.createChecksum;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

import io.github.msgpact.MsgPactSchema;
import io.github.msgpact.internal.types.VArray;
import io.github.msgpact.internal.types.VFn;
import io.github.msgpact.internal.types.VObject;
import io.github.msgpact.internal.types.VStruct;
import io.github.msgpact.internal.types.VTypeDeclaration;
import io.github.msgpact.internal.types.VUnion;

public class ConstructBinaryEncoding {

    private static List<String> traceType(VTypeDeclaration typeDeclaration) {
        final var thisAllKeys = new ArrayList<String>();

        if (typeDeclaration.type instanceof VArray) {
            final var theseKeys2 = traceType(typeDeclaration.typeParameters.get(0));
            thisAllKeys.addAll(theseKeys2);
        } else if (typeDeclaration.type instanceof VObject) {
            final var theseKeys2 = traceType(typeDeclaration.typeParameters.get(0));
            thisAllKeys.addAll(theseKeys2);
        } else if (typeDeclaration.type instanceof VStruct s) {
            final var structFields = s.fields;
            for (final var entry : structFields.entrySet()) {
                final var structFieldKey = entry.getKey();
                final var structField = entry.getValue();
                thisAllKeys.add(structFieldKey);
                final var moreKeys = traceType(structField.typeDeclaration);
                thisAllKeys.addAll(moreKeys);
            }
        } else if (typeDeclaration.type instanceof VUnion u) {
            final var unionTags = u.tags;
            for (final var entry : unionTags.entrySet()) {
                final var tagKey = entry.getKey();
                final var tagValue = entry.getValue();
                thisAllKeys.add(tagKey);
                final var structFields = tagValue.fields;
                for (final var fieldEntry : structFields.entrySet()) {
                    final var structFieldKey = fieldEntry.getKey();
                    final var structField = fieldEntry.getValue();
                    thisAllKeys.add(structFieldKey);
                    final var moreKeys = traceType(structField.typeDeclaration);
                    thisAllKeys.addAll(moreKeys);
                }
            }
        }

        return thisAllKeys;
    }

    public static BinaryEncoding constructBinaryEncoding(MsgPactSchema msgPactSchema) {
        final var allKeys = new TreeSet<String>();
        final var functions = new ArrayList<Map.Entry<String, VFn>>();

        for (final var entry : msgPactSchema.parsed.entrySet()) {
            final var key = entry.getKey();
            final var value = entry.getValue();
            if (value instanceof VFn) {
                functions.add(Map.entry(key, (VFn) value));
            }
        }

        for (final var function : functions) {
            final var key = function.getKey();
            final var value = function.getValue();
            allKeys.add(key);
            final var args = value.call.tags.get(key);
            for (final var fieldEntry : args.fields.entrySet()) {
                final var fieldKey = fieldEntry.getKey();
                final var field = fieldEntry.getValue();
                allKeys.add(fieldKey);
                final var keys = traceType(field.typeDeclaration);
                keys.forEach(allKeys::add);
            }

            final var result = value.result.tags.get("Ok_");
            allKeys.add("Ok_");
            for (final var fieldEntry : result.fields.entrySet()) {
                final var fieldKey = fieldEntry.getKey();
                final var field = fieldEntry.getValue();
                allKeys.add(fieldKey);
                final var keys = traceType(field.typeDeclaration);
                keys.forEach(allKeys::add);
            }
        }

        final var sortedAllKeys = new ArrayList<>(allKeys);
        Collections.sort(sortedAllKeys);

        final var binaryEncoding = new HashMap<String, Integer>();
        for (int index = 0; index < sortedAllKeys.size(); index++) {
            final var key = sortedAllKeys.get(index);
            binaryEncoding.put(key, index);
        }

        final var finalString = String.join("\n", sortedAllKeys);
        final var checksum = createChecksum(finalString);

        return new BinaryEncoding(binaryEncoding, checksum);
    }
}

File: lib/py/msgpact/internal/binary/ConstructBinaryEncoding.py
from typing import List, Dict, Tuple, Set, Union
from typing import TYPE_CHECKING


from ...internal.binary.BinaryEncoding import BinaryEncoding
from ...internal.binary.CreateChecksum import create_checksum

if TYPE_CHECKING:
    from ...MsgPactSchema import MsgPactSchema
    from ..types.VTypeDeclaration import VTypeDeclaration


def trace_type(type_declaration: 'VTypeDeclaration') -> list[str]:
    from ..types.VArray import VArray
    from ..types.VObject import VObject
    from ..types.VStruct import VStruct
    from ..types.VUnion import VUnion

    this_all_keys: list[str] = []

    if isinstance(type_declaration.type, VArray):
        these_keys2 = trace_type(type_declaration.type_parameters[0])
        this_all_keys.extend(these_keys2)
    elif isinstance(type_declaration.type, VObject):
        these_keys2 = trace_type(type_declaration.type_parameters[0])
        this_all_keys.extend(these_keys2)
    elif isinstance(type_declaration.type, VStruct):
        struct_fields = type_declaration.type.fields
        for struct_field_key, struct_field in struct_fields.items():
            this_all_keys.append(struct_field_key)
            more_keys = trace_type(struct_field.type_declaration)
            this_all_keys.extend(more_keys)
    elif isinstance(type_declaration.type, VUnion):
        union_tags = type_declaration.type.tags
        for tag_key, tag_value in union_tags.items():
            this_all_keys.append(tag_key)
            struct_fields = tag_value.fields
            for struct_field_key, struct_field in struct_fields.items():
                this_all_keys.append(struct_field_key)
                more_keys = trace_type(struct_field.type_declaration)
                this_all_keys.extend(more_keys)

    return this_all_keys


def construct_binary_encoding(u_api_schema: 'MsgPactSchema') -> 'BinaryEncoding':
    from ..types.VTypeDeclaration import VTypeDeclaration
    from ..types.VFn import VFn

    all_keys: set[str] = set()

    functions: list[Tuple[str, VFn]] = []

    for key, value in u_api_schema.parsed.items():
        if isinstance(value, VFn):
            functions.append((key, value))

    for key, value in functions:
        all_keys.add(key)
        args = value.call.tags[key]
        for field_key, field in args.fields.items():
            all_keys.add(field_key)
            keys = trace_type(field.type_declaration)
            for key in keys:
                all_keys.add(key)

        result = value.result.tags['Ok_']
        all_keys.add('Ok_')
        for field_key, field in result.fields.items():
            all_keys.add(field_key)
            keys = trace_type(field.type_declaration)
            for key in keys:
                all_keys.add(key)

    sorted_all_keys = sorted(all_keys)

    binary_encoding = {key: i for i, key in enumerate(sorted_all_keys)}
    final_string = "\n".join(sorted_all_keys)
    checksum = create_checksum(final_string)
    return BinaryEncoding(binary_encoding, checksum)
